File: go/engine.go
// Package statusrules is the shipment status transition rule engine.
//
// It answers one question: given a shipment's current status and a proposed
// next status, is this transition allowed by the spec?
//
// All validation is done in-process via an immutable adjacency map compiled
// from the shared rule spec at startup. There are no network calls on the
// hot path. Validate() completes in < 100µs p99 and allocates nothing on
// the success path.
//
// Usage:
//
//	engine := statusrules.NewEngine()
//	result := engine.Validate(current, next, statusrules.FlowForward, ctx)
//	if !result.Valid { ... }
package statusrules

import (
	"os"
	"path/filepath"
	"runtime"
	"sync/atomic"

	"github.com/logistics-oss/status-rules/status"
)

// Pre-defined AnomalyReason strings. String literals live in the data segment —
// assigning them to the result struct never allocates on the heap.
const (
	reasonZeroValue       = "zero-value status: use ValidateInitial for new shipments"
	reasonUnknownStatus   = "unknown status code"
	reasonTerminal        = "transition from a terminal status is never valid"
	reasonInvalidFlow     = "flow must be FlowForward or FlowReverse"
	reasonNotInSpec       = "transition not defined in rule spec"
	reasonMissingOperator = "operator_id required for FORCE_* override transition"
	reasonWrongOwner      = "ctx.SourceSystem does not own the next status"
)

// Engine is the compiled rule engine. It is immutable after construction and
// safe for concurrent use by any number of goroutines without locking.
type Engine struct {
	adj         adjacencyMap
	owners      ownerArray
	terminals   terminalArray
	initials    []initialEntry
	specVersion string

	// Atomic counters — LOCK XADD, no allocation, no contention.
	// Read via Snapshot() from a background goroutine; never on the hot path.
	cntTotal       atomic.Uint64
	cntInvalid     atomic.Uint64
	cntOverride    atomic.Uint64
	cntInterSystem atomic.Uint64
	cntZeroValue   atomic.Uint64
}

// NewEngine initializes the rule engine from the shared spec/rules.yaml.
// Called once at application startup. Panics if the spec cannot be read or
// is structurally invalid.
//
// Initialization time: ~5ms (adjacency map compile + wildcard expansion).
// The returned *Engine is immutable and safe for concurrent use.
func NewEngine() *Engine {
	_, file, _, _ := runtime.Caller(0)
	specPath := filepath.Join(filepath.Dir(file), "..", "spec", "rules.yaml")
	data, err := os.ReadFile(specPath)
	if err != nil {
		panic("status-rules: cannot read spec/rules.yaml: " + err.Error())
	}
	e, err := compileSpec(data)
	if err != nil {
		panic("status-rules: spec is invalid: " + err.Error())
	}
	return e
}

// NewEngineFromSpec initializes from an externally provided spec YAML.
// Used in tests and for canary validation of new rule spec versions before release.
// Not for production initialization.
func NewEngineFromSpec(specYAML []byte) (*Engine, error) {
	return compileSpec(specYAML)
}

// Validate checks whether transitioning from currentStatus to nextStatus
// is permitted in the given flow and context.
//
// This is the hot path. It completes in < 100µs p99.
// It is safe to call from multiple goroutines concurrently.
// It allocates nothing on the heap for the success path.
//
// currentStatus: the shipment's last confirmed status known to the caller.
//
//	The caller is responsible for providing this correctly.
//	The library does not store or retrieve shipment state.
//	Passing a zero-value Status returns ErrZeroValueStatus immediately.
//
// nextStatus: the status the source system intends to emit.
//
//	Passing a zero-value Status returns ErrZeroValueStatus immediately.
//
// flow: the current shipment flow (FlowForward or FlowReverse).
// ctx:  auxiliary metadata — source system identity, operator ID for overrides.
func (e *Engine) Validate(
	currentStatus status.Status,
	nextStatus status.Status,
	flow Flow,
	ctx TransitionContext,
) ValidationResult {
	e.cntTotal.Add(1)

	// ── Guard: zero-value inputs ──────────────────────────────────────────────
	if currentStatus == 0 || nextStatus == 0 {
		e.cntZeroValue.Add(1)
		return ValidationResult{
			ErrorCode:     ErrZeroValueStatus,
			AnomalyReason: reasonZeroValue,
		}
	}

	// ── Guard: unknown status IDs ─────────────────────────────────────────────
	if currentStatus > status.MaxID || nextStatus > status.MaxID {
		e.cntInvalid.Add(1)
		return ValidationResult{
			ErrorCode:     ErrUnknownStatus,
			AnomalyReason: reasonUnknownStatus,
		}
	}

	// ── Guard: valid flow ─────────────────────────────────────────────────────
	if flow != FlowForward && flow != FlowReverse {
		e.cntInvalid.Add(1)
		return ValidationResult{
			ErrorCode:     ErrFlowMismatch,
			AnomalyReason: reasonInvalidFlow,
		}
	}

	// ── Guard: terminal source status ─────────────────────────────────────────
	if e.terminals[currentStatus] {
		e.cntInvalid.Add(1)
		return ValidationResult{
			ErrorCode:     ErrTerminalStatus,
			AnomalyReason: reasonTerminal,
		}
	}

	// ── Hot path: single array lookup ────────────────────────────────────────
	edge := e.adj[packKey(currentStatus, nextStatus, flow)]

	if edge&flagPresent == 0 {
		e.cntInvalid.Add(1)
		return ValidationResult{
			ErrorCode:     ErrInvalidTransition,
			AnomalyReason: reasonNotInSpec,
		}
	}

	// ── Override: operator_id required ────────────────────────────────────────
	if edge&flagRequiresOpID != 0 && ctx.OperatorID == "" {
		e.cntInvalid.Add(1)
		return ValidationResult{
			ErrorCode:     ErrMissingOperatorID,
			AnomalyReason: reasonMissingOperator,
		}
	}

	// ── Inter-system boundary check ───────────────────────────────────────────
	// The emitting system must own the next status.
	isInterSystem := edge&flagInterSystem != 0
	if isInterSystem {
		if e.owners[nextStatus] != ctx.SourceSystem {
			e.cntInvalid.Add(1)
			return ValidationResult{
				ErrorCode:     ErrInvalidSourceSystem,
				AnomalyReason: reasonWrongOwner,
			}
		}
		e.cntInterSystem.Add(1)
	}

	// ── Success path ──────────────────────────────────────────────────────────
	isOverride := edge&flagOverride != 0
	if isOverride {
		e.cntOverride.Add(1)
	}

	mode := ModeIntraSystem
	if isInterSystem {
		mode = ModeInterSystem
	}

	return ValidationResult{
		Valid:      true,
		IsOverride: isOverride,
		Mode:       mode,
		ErrorCode:  ErrNone,
		// AnomalyReason is "" (zero value) — no allocation.
	}
}

// ValidateInitial checks whether a status is a valid first event for a new
// shipment in the given flow. Used by the Booking System on shipment creation,
// where there is no prior status and Validate() must not be called.
func (e *Engine) ValidateInitial(
	firstStatus status.Status,
	flow Flow,
	ctx TransitionContext,
) ValidationResult {
	e.cntTotal.Add(1)

	if firstStatus == 0 {
		e.cntZeroValue.Add(1)
		return ValidationResult{
			ErrorCode:     ErrZeroValueStatus,
			AnomalyReason: reasonZeroValue,
		}
	}
	if firstStatus > status.MaxID {
		e.cntInvalid.Add(1)
		return ValidationResult{
			ErrorCode:     ErrUnknownStatus,
			AnomalyReason: reasonUnknownStatus,
		}
	}
	if flow != FlowForward && flow != FlowReverse {
		e.cntInvalid.Add(1)
		return ValidationResult{
			ErrorCode:     ErrFlowMismatch,
			AnomalyReason: reasonInvalidFlow,
		}
	}

	for _, init := range e.initials {
		if init.statusID == firstStatus && init.flow == flow {
			return ValidationResult{
				Valid:      true,
				Mode:       ModeIntraSystem,
				ErrorCode:  ErrNone,
			}
		}
	}

	e.cntInvalid.Add(1)
	return ValidationResult{
		ErrorCode:     ErrInvalidTransition,
		AnomalyReason: reasonNotInSpec,
	}
}

// PermittedTransitions returns all valid next statuses reachable from
// currentStatus in the given flow. Used by source system UIs to constrain
// operator-facing action menus. Not on the hot path — allocates a slice.
func (e *Engine) PermittedTransitions(
	currentStatus status.Status,
	flow Flow,
) []status.Status {
	if currentStatus == 0 || currentStatus > status.MaxID {
		return nil
	}
	if e.terminals[currentStatus] {
		return nil
	}
	if flow != FlowForward && flow != FlowReverse {
		return nil
	}

	flowBit := uint16(flow - 1)
	var result []status.Status
	for next := status.Status(1); next <= status.MaxID; next++ {
		key := uint16(currentStatus)<<7 | uint16(next)<<1 | flowBit
		if e.adj[key]&flagPresent != 0 {
			result = append(result, next)
		}
	}
	return result
}

// SpecVersion returns the semantic version of the rule spec compiled into this
// engine instance. Source systems must log this on startup and expose it via
// their /health endpoint to enable spec version skew detection.
func (e *Engine) SpecVersion() string {
	return e.specVersion
}

// Snapshot returns a point-in-time copy of all internal counters.
// Call this from your metrics export goroutine (e.g., every 15 seconds).
// This is the only metrics interaction — there are no callbacks on Validate().
func (e *Engine) Snapshot() MetricsSnapshot {
	return MetricsSnapshot{
		ValidationsTotal:       e.cntTotal.Load(),
		InvalidTransitions:     e.cntInvalid.Load(),
		OverrideTransitions:    e.cntOverride.Load(),
		InterSystemTransitions: e.cntInterSystem.Load(),
		ZeroValueRejections:    e.cntZeroValue.Load(),
	}
}

File: node/src/engine.js
'use strict';

/**
 * @fileoverview Status rule engine.
 *
 * Compiles the YAML spec into a flat Uint16Array adjacency table and exposes
 * validate / validateInitial / permittedTransitions helpers.  The design
 * mirrors the Go implementation exactly so that both can be used as reference
 * implementations for compatibility testing.
 *
 * Key packing (matches Go):
 *   key = (fromId << 7) | (toId << 1) | flowBit
 *   flowBit: FORWARD=0, REVERSE=1
 *   Array size: 8192 (covers ids 0-63 × 0-63 × 2 flows)
 *
 * Edge flags (uint16 bit positions):
 *   bit 0 (1):  flagPresent        — transition exists
 *   bit 1 (2):  flagOverride       — FORCE_* edge
 *   bit 2 (4):  flagRequiresOpID   — needs operatorId in context
 *   bit 3 (8):  flagInterSystem    — crosses system ownership boundary
 */

const path = require('path');
const fs   = require('fs');
const yaml = require('js-yaml');

// ── Bit-flag constants ────────────────────────────────────────────────────────

const FLAG_PRESENT         = 1;  // bit 0
const FLAG_OVERRIDE        = 2;  // bit 1
const FLAG_REQUIRES_OP_ID  = 4;  // bit 2
const FLAG_INTER_SYSTEM    = 8;  // bit 3

// ── Adjacency table size ──────────────────────────────────────────────────────

const ADJ_SIZE = 8192; // 64 × 64 × 2

// ── Error codes ───────────────────────────────────────────────────────────────

const ERR_NONE                       = 'ErrNone';
const ERR_ZERO_VALUE                 = 'ErrZeroValueStatus';
const ERR_UNKNOWN_STATUS             = 'ErrUnknownStatus';
const ERR_TERMINAL_STATUS            = 'ErrTerminalStatus';
const ERR_FLOW_MISMATCH              = 'ErrFlowMismatch';
const ERR_INVALID_TRANSITION         = 'ErrInvalidTransition';
const ERR_MISSING_OP_ID              = 'ErrMissingOperatorID';
const ERR_INVALID_SRC_SYS            = 'ErrInvalidSourceSystem';
const ERR_FIRST_MILE_AFTER_LATER_MILE = 'ErrFirstMileAfterLaterMile';

// ── Internal: spec compiler ───────────────────────────────────────────────────

/**
 * Compiled spec data produced by compileSpec().
 *
 * @typedef {Object} CompiledSpec
 * @property {Uint16Array}        adj         - Adjacency table (8192 slots).
 * @property {Map<string,number>} codeToId    - YAML code → numeric id.
 * @property {string[]}           owners      - owners[id] = owned_by string.
 * @property {boolean[]}          terminals   - terminals[id] = true if terminal.
 * @property {Set<string>}        initials    - "code:FLOW" strings for valid first statuses.
 * @property {string}             specVersion - Parsed spec_version field.
 */

/**
 * Compile a YAML spec string into an adjacency table and supporting lookup
 * structures.  This is a pure function — it does not read files or produce
 * side-effects.
 *
 * @param {string} specText - Raw YAML content.
 * @returns {CompiledSpec}
 */
function compileSpec(specText) {
  /** @type {any} */
  const doc = yaml.load(specText);

  const specVersion = String(doc.spec_version || '');

  // ── Step 1: build codeToId (YAML code → yamlId+1) ────────────────────────
  // Slot count: YAML ids are 0-54 → JS ids 1-55.  We allocate 56 slots so
  // that index 55 is valid and index 0 stays the zero sentinel.
  const SLOT_COUNT = 56;

  /** @type {Map<string, number>} */
  const codeToId = new Map();

  /** @type {string[]}  owners[id] = owned_by */
  const owners = new Array(SLOT_COUNT).fill('');

  /** @type {boolean[]} terminals[id] */
  const terminals = new Array(SLOT_COUNT).fill(false);

  /** @type {boolean[]} firstMile[id] — true for FIRST_MILE category statuses */
  const firstMile = new Array(SLOT_COUNT).fill(false);

  /** @type {boolean[]} postFirstMile[id] — true for MIDDLE_MILE or LAST_MILE category statuses */
  const postFirstMile = new Array(SLOT_COUNT).fill(false);

  for (const s of doc.statuses) {
    const id = s.id + 1; // shift: YAML 0-indexed → JS 1-indexed
    codeToId.set(s.code, id);
    owners[id]    = s.owned_by || '';
    terminals[id] = s.terminal === true;
    const cat = (s.category || '').toUpperCase();
    if (cat === 'FIRST_MILE') {
      firstMile[id] = true;
    } else if (cat === 'MIDDLE_MILE' || cat === 'LAST_MILE') {
      postFirstMile[id] = true;
    }
  }

  // ── Step 2: adjacency table ───────────────────────────────────────────────
  const adj = new Uint16Array(ADJ_SIZE);

  /**
   * Pack fromId, toId and flowBit into a table index.
   * @param {number} from
   * @param {number} to
   * @param {number} flowBit - 0=FORWARD, 1=REVERSE
   * @returns {number}
   */
  function key(from, to, flowBit) {
    return (from << 7) | (to << 1) | flowBit;
  }

  /**
   * Return true when fromId and toId belong to different owning systems.
   * @param {number} fromId
   * @param {number} toId
   * @returns {boolean}
   */
  function isInterSystem(fromId, toId) {
    return owners[fromId] !== owners[toId];
  }

  // ── Step 3: expand wildcard targets ──────────────────────────────────────
  // For every non-terminal from-status and every wildcard to-status, in both
  // flows, write the base edge (flagPresent | optionally flagInterSystem).
  // Explicit transitions written in Step 4 will overwrite these entries.

  const wildcardTargets = /** @type {string[]} */ (doc.wildcard_targets || []);

  for (const s of doc.statuses) {
    const fromId = s.id + 1;
    if (terminals[fromId]) continue; // terminal statuses cannot be a source

    for (const wCode of wildcardTargets) {
      const toId = codeToId.get(wCode);
      if (toId === undefined) continue;

      for (const flowBit of [0, 1]) {
        const flags = FLAG_PRESENT | (isInterSystem(fromId, toId) ? FLAG_INTER_SYSTEM : 0);
        adj[key(fromId, toId, flowBit)] = flags;
      }
    }
  }

  // ── Step 4: apply explicit transitions (overwrites wildcard entries) ──────

  for (const t of doc.transitions) {
    const fromId = codeToId.get(t.from);
    if (fromId === undefined) continue;

    const flowBit = t.flow === 'REVERSE' ? 1 : 0;
    const isOverride    = t.is_override === true;
    const requiresOpId  = Array.isArray(t.requires_context) &&
                          t.requires_context.includes('operator_id');

    // `to` may be a single string or an array
    const targets = Array.isArray(t.to) ? t.to : [t.to];

    for (const toCode of targets) {
      const toId = codeToId.get(toCode);
      if (toId === undefined) continue;

      const flags =
        FLAG_PRESENT |
        (isOverride   ? FLAG_OVERRIDE       : 0) |
        (requiresOpId ? FLAG_REQUIRES_OP_ID : 0) |
        (isInterSystem(fromId, toId) ? FLAG_INTER_SYSTEM : 0);

      adj[key(fromId, toId, flowBit)] = flags;
    }
  }

  // ── Step 5: build initials set ────────────────────────────────────────────
  /** @type {Set<string>} */
  const initials = new Set();
  for (const entry of (doc.initial_statuses || [])) {
    initials.add(`${entry.code}:${entry.flow}`);
  }

  return { adj, codeToId, owners, terminals, firstMile, postFirstMile, initials, specVersion };
}

// ── Engine class ──────────────────────────────────────────────────────────────

/**
 * Validation context passed to validate().
 *
 * @typedef {Object} ValidationContext
 * @property {string} [operatorId]   - Required for FORCE_* transitions.
 * @property {string} [sourceSystem] - Caller's system identity; validated on
 *                                     inter-system edges.
 */

/**
 * Result returned by validate() and validateInitial().
 *
 * @typedef {Object} ValidationResult
 * @property {boolean} valid          - Whether the transition is permitted.
 * @property {boolean} isOverride     - True for FORCE_* override edges.
 * @property {boolean} isInterSystem  - True when the edge crosses an ownership
 *                                      boundary between source systems.
 * @property {string}  errorCode      - One of the ErrXxx constants; 'ErrNone'
 *                                      on success.
 */

/**
 * @typedef {Object} PermittedTransition
 * @property {string}  code         - Destination status code (YAML name).
 * @property {boolean} isOverride
 * @property {boolean} isInterSystem
 * @property {boolean} requiresOperatorId
 */

/**
 * Snapshot of engine metrics returned by snapshot().
 *
 * @typedef {Object} EngineSnapshot
 * @property {string} specVersion
 * @property {number} cntTotal
 * @property {number} cntInvalid
 * @property {number} cntOverride
 * @property {number} cntInterSystem
 * @property {number} cntZeroValue
 */

/**
 * Shipment status rule engine.
 *
 * Reads the YAML spec at construction time and compiles it into a dense
 * adjacency table for O(1) lookups.
 *
 * @example
 * const { Engine, Flow, SourceSystem } = require('@logistics-oss/status-rules');
 * const engine = new Engine();
 * const result = engine.validate('HUB-IN-SCANNED', 'HUB-OUTSCAN', Flow.FORWARD, {});
 */
class Engine {
  constructor() {
    // Resolve the spec path relative to this source file so the package works
    // regardless of the working directory from which it is imported.
    const specPath = path.resolve(__dirname, '../../spec/rules.yaml');
    const specText = fs.readFileSync(specPath, 'utf8');

    /** @type {CompiledSpec} */
    this._spec = compileSpec(specText);

    // ── Metrics counters ──────────────────────────────────────────────────
    /** @type {number} Total calls to validate() */
    this.cntTotal        = 0;
    /** @type {number} Calls that returned valid=false */
    this.cntInvalid      = 0;
    /** @type {number} Successful calls where isOverride=true */
    this.cntOverride     = 0;
    /** @type {number} Successful calls where isInterSystem=true */
    this.cntInterSystem  = 0;
    /** @type {number} Calls that returned ErrZeroValueStatus */
    this.cntZeroValue    = 0;
  }

  // ── Public API ────────────────────────────────────────────────────────────

  /**
   * Spec version string from the YAML (e.g. "1.0.0").
   * @returns {string}
   */
  get specVersion() {
    return this._spec.specVersion;
  }

  /**
   * Validate a status transition.
   *
   * Follows the algorithm specified in the architecture document, matching the
   * Go implementation step-for-step.
   *
   * @param {string}            current - Current status code (YAML name).
   * @param {string}            next    - Desired next status code (YAML name).
   * @param {string}            flow    - 'FORWARD' or 'REVERSE'.
   * @param {ValidationContext} ctx     - Caller-supplied context.
   * @returns {ValidationResult}
   */
  validate(current, next, flow, ctx) {
    this.cntTotal++;

    // ── Guard: zero-value inputs ──────────────────────────────────────────
    if (!current || !next) {
      this.cntInvalid++;
      this.cntZeroValue++;
      return { valid: false, isOverride: false, isInterSystem: false, errorCode: ERR_ZERO_VALUE };
    }

    const { adj, codeToId, owners, terminals, firstMile, postFirstMile } = this._spec;

    // ── Guard: unknown status codes ───────────────────────────────────────
    const currentId = codeToId.get(current);
    const nextId    = codeToId.get(next);

    if (currentId === undefined || nextId === undefined) {
      this.cntInvalid++;
      return { valid: false, isOverride: false, isInterSystem: false, errorCode: ERR_UNKNOWN_STATUS };
    }

    // ── Guard: current status is terminal ────────────────────────────────
    if (terminals[currentId]) {
      this.cntInvalid++;
      return { valid: false, isOverride: false, isInterSystem: false, errorCode: ERR_TERMINAL_STATUS };
    }

    // ── Guard: first-mile status cannot follow middle-mile or last-mile ───
    if (postFirstMile[currentId] && firstMile[nextId]) {
      this.cntInvalid++;
      return { valid: false, isOverride: false, isInterSystem: false, errorCode: ERR_FIRST_MILE_AFTER_LATER_MILE };
    }

    // ── Look up edge in adjacency table ───────────────────────────────────
    const flowBit    = flow === 'FORWARD' ? 0 : 1;
    const edgeKey    = (currentId << 7) | (nextId << 1) | flowBit;
    const edge       = adj[edgeKey];

    if (!(edge & FLAG_PRESENT)) {
      // Edge not found in the requested flow — check if it exists in the
      // opposite flow to distinguish ErrFlowMismatch from ErrInvalidTransition.
      const altFlowBit = flowBit ^ 1;
      const altKey     = (currentId << 7) | (nextId << 1) | altFlowBit;
      const altEdge    = adj[altKey];

      this.cntInvalid++;
      if (altEdge & FLAG_PRESENT) {
        return { valid: false, isOverride: false, isInterSystem: false, errorCode: ERR_FLOW_MISMATCH };
      }
      return { valid: false, isOverride: false, isInterSystem: false, errorCode: ERR_INVALID_TRANSITION };
    }

    // ── Guard: operator id required ───────────────────────────────────────
    if ((edge & FLAG_REQUIRES_OP_ID) && !ctx.operatorId) {
      this.cntInvalid++;
      return { valid: false, isOverride: false, isInterSystem: false, errorCode: ERR_MISSING_OP_ID };
    }

    // ── Guard: inter-system ownership check ───────────────────────────────
    const interSystem = !!(edge & FLAG_INTER_SYSTEM);
    if (interSystem && owners[nextId] !== ctx.sourceSystem) {
      this.cntInvalid++;
      return { valid: false, isOverride: false, isInterSystem: true, errorCode: ERR_INVALID_SRC_SYS };
    }

    // ── Success ───────────────────────────────────────────────────────────
    const override = !!(edge & FLAG_OVERRIDE);
    if (override)     this.cntOverride++;
    if (interSystem)  this.cntInterSystem++;

    return { valid: true, isOverride: override, isInterSystem: interSystem, errorCode: ERR_NONE };
  }

  /**
   * Validate the very first status of a new shipment.
   *
   * Unlike validate(), there is no prior status to check against — this method
   * consults the `initial_statuses` list in the spec.
   *
   * @param {string} first - The first status code.
   * @param {string} flow  - 'FORWARD' or 'REVERSE'.
   * @returns {ValidationResult}
   */
  validateInitial(first, flow) {
    if (!first) {
      return { valid: false, isOverride: false, isInterSystem: false, errorCode: ERR_ZERO_VALUE };
    }

    if (!this._spec.codeToId.has(first)) {
      return { valid: false, isOverride: false, isInterSystem: false, errorCode: ERR_UNKNOWN_STATUS };
    }

    const key = `${first}:${flow}`;
    if (this._spec.initials.has(key)) {
      return { valid: true, isOverride: false, isInterSystem: false, errorCode: ERR_NONE };
    }

    return { valid: false, isOverride: false, isInterSystem: false, errorCode: ERR_INVALID_TRANSITION };
  }

  /**
   * Return all permitted next statuses reachable from `current` in `flow`.
   *
   * @param {string} current - Current status code (YAML name).
   * @param {string} flow    - 'FORWARD' or 'REVERSE'.
   * @returns {PermittedTransition[]}
   */
  permittedTransitions(current, flow) {
    const { adj, codeToId, terminals } = this._spec;
    const { STATUS_NAMES } = require('./status');

    const currentId = codeToId.get(current);
    if (currentId === undefined || terminals[currentId]) return [];

    const flowBit = flow === 'FORWARD' ? 0 : 1;
    const results = [];

    // Iterate over all possible destination ids (1 … MAX_ID).
    const maxId = STATUS_NAMES.length - 1; // STATUS_NAMES[0] is the sentinel
    for (let toId = 1; toId <= maxId; toId++) {
      const edgeKey = (currentId << 7) | (toId << 1) | flowBit;
      const edge    = adj[edgeKey];
      if (!(edge & FLAG_PRESENT)) continue;

      results.push({
        code:               STATUS_NAMES[toId],
        isOverride:         !!(edge & FLAG_OVERRIDE),
        isInterSystem:      !!(edge & FLAG_INTER_SYSTEM),
        requiresOperatorId: !!(edge & FLAG_REQUIRES_OP_ID),
      });
    }

    return results;
  }

  /**
   * Return a point-in-time snapshot of engine metrics and spec metadata.
   *
   * @returns {EngineSnapshot}
   */
  snapshot() {
    return {
      specVersion:    this._spec.specVersion,
      cntTotal:       this.cntTotal,
      cntInvalid:     this.cntInvalid,
      cntOverride:    this.cntOverride,
      cntInterSystem: this.cntInterSystem,
      cntZeroValue:   this.cntZeroValue,
    };
  }
}

module.exports = { Engine };
